### ISSUE TITLE
#142 transferSAFEOwnership() does not fully transfer ownership 

### ISSUE THREAD

**Issue by pi0neerpat** [open-dollar/od-contracts#217]
> # Lines of code
> https://github.com/open-dollar/od-contracts/blob/f4f0246bb26277249c1d5afe6201d4d9096e52e6/src/contracts/proxies/ODSafeManager.sol#L136-L152
> 
> # Vulnerability details
> ## Impact
> Function `transferSAFEOwnership()` is only callable by `Only Vault721`: `require(msg.sender == address(vault721), 'SafeMngr: Only Vault721');` and its purpose is to give the safe ownership to a `dst` address. However, `transferSAFEOwnership()` does not reset `safeCan` mapping which is used by `safeAllowed` modifier.
> 
> This leads to some security issues. The modifier `safeAllowed` is used to verify if user has permission to the safe. Multiple of functions which perform critical operations on the safe use this modifier to verify is user has permission to call that operation on a given safe.
> 
> Since `transferSAFEOwnership()` transfer ownership to a different address, it should also reset all previously set privileges by the previous owner. However, `safeCan` mapping is not being reset in `transferSAFEOwnership()`
> 
> ## Proof of Concept
> Firstly, let's take a look how `safeAllowed` modifier is defined:
> 
> [File: ODSafeManager.sol](https://github.com/open-dollar/od-contracts/blob/f4f0246bb26277249c1d5afe6201d4d9096e52e6/src/contracts/proxies/ODSafeManager.sol#L49-L53)
> 
> ```
>   modifier safeAllowed(uint256 _safe) {
>     address _owner = _safeData[_safe].owner;
>     if (msg.sender != _owner && safeCan[_owner][_safe][msg.sender] == 0) revert SafeNotAllowed();
>     _;
>   }
> ```
> 
> For this PoC, let's consider user `A`, who's the owner of safe 123. He gives access to the safe to `X`, `Y`, so he calls: `allowSAFE(123, X, 1), allowSAFE(123, Y, 1)`;
> 
> [File: ODSafeManager.sol](https://github.com/open-dollar/od-contracts/blob/f4f0246bb26277249c1d5afe6201d4d9096e52e6/src/contracts/proxies/ODSafeManager.sol#L49-L53)
> 
> ```
>  function allowSAFE(uint256 _safe, address _usr, uint256 _ok) external safeAllowed(_safe) {
>     address _owner = _safeData[_safe].owner;
>     safeCan[_owner][_safe][_usr] = _ok;
>     emit AllowSAFE(msg.sender, _safe, _usr, _ok);
>   }
> ```
> 
> After calling this function, `safeCan` mapping looks like this: `safeCan[A][123][X] = 1` and `safeCan[A][123][Y] = 1`.
> 
> Now, there's a call to transfer safe ownership to `_dst` B: `transferSAFEOwnership(123, B)`
> 
> [File: ODSafeManager.sol](https://github.com/open-dollar/od-contracts/blob/f4f0246bb26277249c1d5afe6201d4d9096e52e6/src/contracts/proxies/ODSafeManager.sol#L49-L53)
> 
> ```
> function transferSAFEOwnership(uint256 _safe, address _dst) external {
>     require(msg.sender == address(vault721), 'SafeMngr: Only Vault721');
> 
>     if (_dst == address(0)) revert ZeroAddress();
>     SAFEData memory _sData = _safeData[_safe];
>     if (_dst == _sData.owner) revert AlreadySafeOwner();
> 
>     _usrSafes[_sData.owner].remove(_safe);
>     _usrSafesPerCollat[_sData.owner][_sData.collateralType].remove(_safe);
> 
>     _usrSafes[_dst].add(_safe);
>     _usrSafesPerCollat[_dst][_sData.collateralType].add(_safe);
> 
>     _safeData[_safe].owner = _dst;
> 
>     emit TransferSAFEOwnership(msg.sender, _safe, _dst);
>   }
> ```
> 
> It modifies only `_usrSafes`, `_usrSafesPerCollat` and `_safeData` mappings and does not modify `safeCan` mapping. After `transferSAFEOwnership(123, B)` call, those mappings look like this:
> 
> ```
> _usrSafes[B] = [123]
> _usrSafesPerCollat[B][_sData.collateralType] = [123]
> _safeData[123].owner = B;
> 
> safeCan[A][123][X]  // notice this is from previous allowSAFE() call
> safeCan[A][123][Y]  // notice this is from previous allowSAFE() call
> ```
> 
> Now, at some point in the future, the safe goes back to user A. Function `transferSAFEOwnership()` is being called once again: `transferSAFEOwnership(123, A)`.
> 
> User `A` get access to the safe. However, he does not remember that he previously sets access to users `X` and `Y`. Those mappings:
> 
> ```
> safeCan[A][123][X]
> safeCan[A][123][Y]
> ```
> 
> are still there - meaning that `X` and `Y` has still access to that safe, even though the safe was just transferred back to `A` and safe's permissions should be cleared.
> 
> Now, when user `X` or `Y` will try to call any function with `safeAllowed` modifier - they will be allowed, since `safeAllowed` will return true for them, because `X` and `Y` are still in `safeCan` mapping.
> 
> ## Tools Used
> Manual code review
> 
> ## Recommended Mitigation Steps
> Whenever transfer ownership - make sure to reset `safeCan[old_owner][safe]` mapping.
> 
> ## Assessed type
> Access Control

**Comment by 0xdavinchee** [comment]
gm!

### High-level Summary
This issue is pointing out that the approvals (`safeCan`) for a safe are not cleared when ownership of a SAFE is transferred. 
The scenario pointed out is this:
- Alice approves Bob and Carol
- Alice transfers the safe to Dave
- Dave transfers the safe back to Alice
- Bob and Carol still have approval (in the `safeCan` mapping) and therefore satisfy the `safeAllowed` modifier and can call any functions which is normally permissioned by this modifier

### Assessment
While this is certainly the case, the recommended mitigation step is inaccurate as you cannot clear the `safeCan` approvals due to the fact that this is a mapping and there is no way of knowing all the users who have approval without knowing the addresses beforehand. 

Saving these addresses in an EnumerableSet or array would not be scalable either during removal. Imagine if Alice decided to add a ludicrously high number of approvals via the `allowSAFE` function. This is subject to the block gas limit limitation and the safe wouldn't be transferable anymore.

It is important to note that this sort of `mapping => mapping` pattern exists elsewhere, for example in ERC721, in the `isApprovedForAll` mapping. 

When a token is transferred, this mapping is not cleared either.

### Possible Solutions
#### On-Chain Solutions
There are certainly on-chain solutions which could address this, one would be adding another nest of mapping which goes from `mapping(uint256 _safeNonce => mapping(address _caller => uint256 _ok))`.

Where you have an idea of a safeNonce which is incremented every time the safe is transferred which would render the previous `safeCan` approvals nil. 

We could add a new `uint96 safeNonce` property to the `SAFEData` struct, we use uint96 so it can be packed before/after `address owner` or `address safeHandler`. uint96 should be a sufficient number of transfers as the max value is: `7.922816251426434e+28`.

#### Off-Chain Solutions
An off-chain solution would be providing some indication of the approvals that have been set in the UI and if there is a subgraph it will index all `AllowSAFE` events emitted.
The UI could warn and prompt the user to clear all or some of the approvals before they are allowed to transfer their SAFE.
Given that each user has their own proxy, this can be batched but is subject to the block gas limit limitation as well.
However, this solution does not add complexity to the smart contracts and moves the complexity to off-chain components like the subgraph and the front-end which is always ideal.


### TODOS
- [ ] Decide on which possible solution to use OR come up with another alternative solution

**Comment by pi0neerpat** [comment]
I like the on-chain solution. There is another project ongoing which deals with nonces to prevent changes to a safe prior to a sale on an NFT marketplace. I think its best that we put this issue on hold for now, as the other issue is a higher priority, and the resulting solution could be used here as well.

**Comment by 0xdavinchee** [comment]
Noted, will hold on this for now.

**Comment by 0xdavinchee** [comment]
currently blocked by #277
